File: app/routes/prescriptions.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
import uuid

from app.database import get_db
from app.models.users import User
from app.models.prescriptions import Prescription, Medication
from app.models.clinics import Clinic
from app.models.patients import Patient
from app.schemas.prescription import (
    PrescriptionCreate,
    PrescriptionResponse,
    PrescriptionUpdate,
    MedicationCreate
)
from app.auth import get_current_active_user

router = APIRouter()

# Get all prescriptions (admin only in a real app)
@router.get("/", response_model=List[PrescriptionResponse])
async def get_prescriptions(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    # In a real app, check if admin
    prescriptions = db.query(Prescription).all()
    
    # Enrich with related data
    result = []
    for prescription in prescriptions:
        clinic = db.query(Clinic).filter(Clinic.id == prescription.clinic_id).first()
        patient = db.query(Patient).filter(Patient.id == prescription.patient_id).first()
        
        prescription_dict = {
            **prescription.__dict__,
            "clinic_name": clinic.name if clinic else None,
            "patient_name": patient.name if patient else None,
            "medications": prescription.medications
        }
        result.append(prescription_dict)
    
    return result

# Create a prescription
@router.post("/", response_model=PrescriptionResponse)
async def create_prescription(
    prescription_data: PrescriptionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    # Check if clinic
    if current_user.type != "clinic":
        raise HTTPException(status_code=403, detail="Only clinics can create prescriptions")
    
    # Check if patient exists
    patient = db.query(Patient).filter(Patient.id == prescription_data.patient_id).first()
    if not patient:
        raise HTTPException(status_code=404, detail="Patient not found")
    
    # Create prescription
    prescription = Prescription(
        id=str(uuid.uuid4()),
        patient_id=prescription_data.patient_id,
        clinic_id=current_user.id,
        issue_date=prescription_data.issue_date,
        valid_until=prescription_data.valid_until,
        notes=prescription_data.notes
    )
    
    db.add(prescription)
    db.flush()  # Flush to get the ID
    
    # Create medications
    medications = []
    for med_data in prescription_data.medications:
        medication = Medication(
            id=str(uuid.uuid4()),
            prescription_id=prescription.id,
            name=med_data.name,
            dosage=med_data.dosage,
            frequency=med_data.frequency,
            duration=med_data.duration,
            notes=med_data.notes
        )
        
        db.add(medication)
        medications.append(medication)
    
    db.commit()
    db.refresh(prescription)
    
    # Return with additional info
    clinic = db.query(Clinic).filter(Clinic.id == prescription.clinic_id).first()
    
    response = {
        **prescription.__dict__,
        "clinic_name": clinic.name if clinic else None,
        "patient_name": patient.name if patient else None,
        "medications": medications
    }
    
    return response

# Get patient prescriptions
@router.get("/patient/{patient_id}", response_model=List[PrescriptionResponse])
async def get_patient_prescriptions(
    patient_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    # Check if user is authorized
    if current_user.id != patient_id and current_user.type != "clinic":
        raise HTTPException(status_code=403, detail="Not authorized to view these prescriptions")
    
    prescriptions = db.query(Prescription).filter(Prescription.patient_id == patient_id).all()
    
    # Enrich with related data
    result = []
    for prescription in prescriptions:
        clinic = db.query(Clinic).filter(Clinic.id == prescription.clinic_id).first()
        patient = db.query(Patient).filter(Patient.id == prescription.patient_id).first()
        
        prescription_dict = {
            **prescription.__dict__,
            "clinic_name": clinic.name if clinic else None,
            "patient_name": patient.name if patient else None,
            "medications": prescription.medications
        }
        result.append(prescription_dict)
    
    return result

# Get clinic prescriptions
@router.get("/clinic/{clinic_id}", response_model=List[PrescriptionResponse])
async def get_clinic_prescriptions(
    clinic_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    # Check if user is authorized
    if current_user.id != clinic_id and current_user.type != "clinic":
        raise HTTPException(status_code=403, detail="Not authorized to view these prescriptions")
    
    prescriptions = db.query(Prescription).filter(Prescription.clinic_id == clinic_id).all()
    
    # Enrich with related data
    result = []
    for prescription in prescriptions:
        patient = db.query(Patient).filter(Patient.id == prescription.patient_id).first()
        
        prescription_dict = {
            **prescription.__dict__,
            "clinic_name": current_user.name,
            "patient_name": patient.name if patient else None,
            "medications": prescription.medications
        }
        result.append(prescription_dict)
    
    return result

# Get a specific prescription
@router.get("/{prescription_id}", response_model=PrescriptionResponse)
async def get_prescription(
    prescription_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    prescription = db.query(Prescription).filter(Prescription.id == prescription_id).first()
    if not prescription:
        raise HTTPException(status_code=404, detail="Prescription not found")
    
    # Check if user is authorized
    if (current_user.id != prescription.patient_id and 
        current_user.id != prescription.clinic_id and 
        current_user.type != "clinic"):
        raise HTTPException(status_code=403, detail="Not authorized to view this prescription")
    
    # Enrich with related data
    clinic = db.query(Clinic).filter(Clinic.id == prescription.clinic_id).first()
    patient = db.query(Patient).filter(Patient.id == prescription.patient_id).first()
    
    response = {
        **prescription.__dict__,
        "clinic_name": clinic.name if clinic else None,
        "patient_name": patient.name if patient else None,
        "medications": prescription.medications
    }
    
    return response

# Update a prescription
@router.put("/{prescription_id}", response_model=PrescriptionResponse)
async def update_prescription(
    prescription_id: str,
    prescription_data: PrescriptionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    # Get prescription
    prescription = db.query(Prescription).filter(Prescription.id == prescription_id).first()
    if not prescription:
        raise HTTPException(status_code=404, detail="Prescription not found")
    
    # Check if user is authorized
    if current_user.id != prescription.clinic_id or current_user.type != "clinic":
        raise HTTPException(status_code=403, detail="Not authorized to update this prescription")
    
    # Update prescription
    for key, value in prescription_data.dict(exclude_unset=True).items():
        setattr(prescription, key, value)
    
    db.commit()
    db.refresh(prescription)
    
    # Enrich with related data
    clinic = db.query(Clinic).filter(Clinic.id == prescription.clinic_id).first()
    patient = db.query(Patient).filter(Patient.id == prescription.patient_id).first()
    
    response = {
        **prescription.__dict__,
        "clinic_name": clinic.name if clinic else None,
        "patient_name": patient.name if patient else None,
        "medications": prescription.medications
    }
    
    return response

File: routes/prescriptions.js
const express = require('express');
const router = express.Router();
const { v4: uuidv4 } = require('uuid');

const { authenticate } = require('./auth');
const { 
  getPrescriptions, 
  getPrescriptionById, 
  addPrescription, 
  getPatientPrescriptions, 
  getClinicPrescriptions 
} = require('../models/prescriptions');
const { getClinicById } = require('../models/clinics');
const { getPatientById } = require('../models/patients');
const { createOrderFromPrescription } = require('../models/products');

// Get patient prescriptions
router.get('/patient/:patientId', authenticate, async (req, res) => {
  try {
    const { patientId } = req.params;
    
    // Check if user is authorized
    if (req.user.id !== patientId && req.user.type !== 'patient') {
      return res.status(403).json({ message: 'Unauthorized to view these prescriptions' });
    }
    
    const prescriptions = getPatientPrescriptions(patientId);
    
    res.json(prescriptions);
  } catch (error) {
    console.error(`Error fetching patient prescriptions ${req.params.patientId}:`, error);
    res.status(500).json({ message: 'Server error while fetching patient prescriptions' });
  }
});

// Get clinic prescriptions
router.get('/clinic/:clinicId', authenticate, async (req, res) => {
  try {
    const { clinicId } = req.params;
    
    // Check if user is authorized
    if (req.user.id !== clinicId && req.user.type !== 'clinic') {
      return res.status(403).json({ message: 'Unauthorized to view these prescriptions' });
    }
    
    const prescriptions = getClinicPrescriptions(clinicId);
    
    res.json(prescriptions);
  } catch (error) {
    console.error(`Error fetching clinic prescriptions ${req.params.clinicId}:`, error);
    res.status(500).json({ message: 'Server error while fetching clinic prescriptions' });
  }
});

// Get prescription by ID
router.get('/:id', authenticate, async (req, res) => {
  try {
    const { id } = req.params;
    const prescription = getPrescriptionById(id);
    
    if (!prescription) {
      return res.status(404).json({ message: 'Prescription not found' });
    }
    
    // Check if user is authorized
    if (req.user.id !== prescription.patientId && req.user.id !== prescription.clinicId) {
      return res.status(403).json({ message: 'Unauthorized to view this prescription' });
    }
    
    res.json(prescription);
  } catch (error) {
    console.error(`Error fetching prescription ${req.params.id}:`, error);
    res.status(500).json({ message: 'Server error while fetching prescription' });
  }
});

// Create new prescription
router.post('/', authenticate, async (req, res) => {
  try {
    const { clinicId, patientId, doctorName, diagnosis, medications, notes } = req.body;
    
    // Check if user is authorized
    if (req.user.id !== clinicId || req.user.type !== 'clinic') {
      return res.status(403).json({ message: 'Unauthorized to create prescriptions for this clinic' });
    }
    
    // Check if clinic exists
    const clinic = getClinicById(clinicId);
    if (!clinic) {
      return res.status(404).json({ message: 'Clinic not found' });
    }
    
    // Check if patient exists
    const patient = getPatientById(patientId);
    if (!patient) {
      return res.status(404).json({ message: 'Patient not found' });
    }
    
    // Create prescription
    const prescription = {
      id: uuidv4(),
      clinicId,
      patientId,
      clinicName: clinic.name,
      patientName: patient.name,
      doctorName,
      diagnosis,
      medications: medications.map(med => ({
        id: uuidv4(),
        name: med.name,
        dosage: med.dosage,
        frequency: med.frequency,
        duration: med.duration
      })),
      notes,
      date: new Date().toISOString()
    };
    
    // Add prescription
    const newPrescription = addPrescription(prescription);
    
    res.status(201).json(newPrescription);
  } catch (error) {
    console.error('Error creating prescription:', error);
    res.status(500).json({ message: 'Server error while creating prescription' });
  }
});

// Order prescription medications
router.post('/order', authenticate, async (req, res) => {
  try {
    const { prescriptionId, patientId, medicationIds } = req.body;
    
    // Check if user is authorized
    if (req.user.id !== patientId || req.user.type !== 'patient') {
      return res.status(403).json({ message: 'Unauthorized to order medications for this patient' });
    }
    
    // Check if prescription exists
    const prescription = getPrescriptionById(prescriptionId);
    if (!prescription) {
      return res.status(404).json({ message: 'Prescription not found' });
    }
    
    // Check if this is the patient's prescription
    if (prescription.patientId !== patientId) {
      return res.status(403).json({ message: 'Unauthorized to order medications from this prescription' });
    }
    
    // Create order from prescription
    const order = createOrderFromPrescription(prescription, medicationIds);
    
    res.status(201).json(order);
  } catch (error) {
    console.error('Error ordering prescription medications:', error);
    res.status(500).json({ message: 'Server error while ordering prescription medications' });
  }
});

module.exports = router;
